tools/cache.py

```diff
@@ -82,30 +83,30 @@ def ensure(self):
 
   def erase(self):
     with self.lock():
-      if os.path.exists(self.dirname):
+      if self.dirname.exists():
         for f in os.listdir(self.dirname):
-          tempfiles.try_delete(os.path.join(self.dirname, f))
+          tempfiles.try_delete(Path(self.dirname, f))
 
   def get_path(self, name):
-    return os.path.join(self.dirname, name)
+    return Path(self.dirname, name)
 
   def get_sysroot(self, absolute):
     if absolute:
       return os.path.join(self.dirname, 'sysroot')
     return 'sysroot'
 
   def get_include_dir(self, *parts):
-    return self.get_sysroot_dir('include', *parts)
+    return str(self.get_sysroot_dir('include', *parts))
 
   def get_sysroot_dir(self, *parts):
-    return os.path.join(self.get_sysroot(absolute=True), *parts)
+    return str(Path(self.get_sysroot(absolute=True), *parts))
```

Comment:
Same question here. I'm just not sure what criteria we have to decide when to use this pattern and when the other?

Reply:
Basically I've converting as much as possible of the internal usage within this file to `Path` and away from `str`.

In order to avoid have that effect a lot more of the codebase (just now) I'm still returning strings at the external API boundary in some cases.   Eventually we would switch away from strings completely have use the higher level sturcutred `Path` type everywhere.  So the choice is just about limited the impact of this incremental change.  

We could draw the boundary anywhere we like and I imagine I will have  followup to convert a lot more of the codebase in the future.

Reply:
Two main advantages:

1.  Allows us to write `'foo/bar/baz.txt'` in the source code and have the be autoamtically converted to a windows path when running on windoes (rather than having to write `os.path.join('foo', 'bar', 'baz.txt')` which is less readable and less searchable/grepable).
2. Its higher level datatype, rather than just a string.. which means we get a little more type checking from python.

Reply:
I think (1) was probably more important for the test code, but as you can see form this change it also helps sometimes in non-test code.